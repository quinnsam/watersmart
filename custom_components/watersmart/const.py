"""Constants for the WaterSmart integration."""

from datetime import timedelta
from enum import StrEnum, auto
from typing import Final

ATTRIBUTION: Final = "Data scraped from WaterSmart"
DOMAIN: Final = "watersmart"
MANUFACTURER: Final = "WaterSmart by VertexOne"
DEFAULT_SCAN_INTERVAL = timedelta(hours=1)


class SensorKey(StrEnum):
    """Converter key enumeration class."""

    GALLONS_FOR_MOST_RECENT_HOUR = auto()
    GALLONS_FOR_MOST_RECENT_FULL_DAY_KEY = auto()
