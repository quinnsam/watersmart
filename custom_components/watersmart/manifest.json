{
  "codeowners": ["@wbyoung"],
  "config_flow": true,
  "dependencies": [],
  "documentation": "https://github.com/wbyoung/watersmart/",
  "domain": "watersmart",
  "iot_class": "cloud_polling",
  "name": "WaterSmart",
  "requirements": [],
  "version": "0.1.0"
}
