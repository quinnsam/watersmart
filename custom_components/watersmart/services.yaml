get_water_usage_hourly_history:
  fields:
    config_entry:
      required: true
      selector:
        config_entry:
          integration: watersmart
    cached:
      required: false
      default: false
      selector:
        boolean:
    start:
      required: false
      example: "2024-01-01 00:00:00"
      selector:
        datetime:
    end:
      required: false
      example: "2024-01-01 00:00:00"
      selector:
        datetime:
