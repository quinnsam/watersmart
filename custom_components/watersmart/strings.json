{
  "config": {
    "step": {
      "user": {
        "data": {
          "host": "[%key:common::config_flow::data::host%]",
          "username": "[%key:common::config_flow::data::username%]",
          "password": "[%key:common::config_flow::data::password%]"
        },
        "description": "Enter your host and credentials. The host is the subdomain for accessing Watersmart, i.e. `bendoregon` for https://bendoregon.watersmart.com/.",
        "title": "Subdomain & Authentication"
      }
    },
    "error": {
      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
      "unknown": "[%key:common::config_flow::error::unknown%]"
    },
    "abort": {
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
    }
  },
  "exceptions": {
    "invalid_date": {
      "message": "Invalid date provided. Got {date}"
    },
    "invalid_config_entry": {
      "message": "Invalid config entry provided. Got {config_entry}"
    },
    "unloaded_config_entry": {
      "message": "Invalid config entry provided. {config_entry} is not loaded."
    }
  },
  "entity": {
    "sensor": {
      "gallons_for_most_recent_hour": {
        "name": "Most recent hour usage"
      },
      "gallons_for_most_recent_full_day": {
        "name": "Most recent full day usage"
      }
    }
  },
  "services": {
    "get_water_usage_hourly_history": {
      "name": "Get hourly water usage history",
      "description": "Request hourly water usage from WaterSmart.",
      "fields": {
        "config_entry": {
          "name": "Config Entry",
          "description": "The config entry to use for this service."
        },
        "cached": {
          "name": "Cached Data",
          "description": "Accept data from the integration cache instead of re-fetching."
        },
        "start": {
          "name": "Start",
          "description": "Specifies the date and time from which to retrieve usage."
        },
        "end": {
          "name": "End",
          "description": "Specifies the date and time until which to retrieve usage."
        }
      }
    }
  }
}
