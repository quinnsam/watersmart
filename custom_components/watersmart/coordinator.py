"""The WaterSmart coordinator."""

from asyncio import timeout
from collections.abc import Callable
import datetime as dt
import functools
import logging
from typing import Any, Protocol, TypedDict, cast

from aiohttp.client_exceptions import ClientConnectorError
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
from homeassistant.util.dt import as_local, get_default_time_zone, start_of_local_day

from .client import AuthenticationError, UsageRecord, WaterSmartClient
from .const import DEFAULT_SCAN_INTERVAL, DOMAIN, MANUFACTURER, SensorKey
from .types import SensorData

EXCEPTIONS = (AuthenticationError, ClientConnectorError)

_LOGGER = logging.getLogger(__name__)


class CoordinatorData(TypedDict, total=False):
    """Shape of coordinator data."""

    gallons_for_most_recent_hour: SensorData
    gallons_for_most_recent_full_day: SensorData
    gallons_today: SensorData
    hourly: list[UsageRecord]


class _DataConverterT(Protocol):
    converter_key: SensorKey

    def __call__(self, data: CoordinatorData) -> SensorData: ...  # pragma no cover


class WaterSmartUpdateCoordinator(DataUpdateCoordinator[CoordinatorData]):
    """Class to manage fetching Watersmart data."""

    data_converters: tuple[_DataConverterT, ...] = ()

    def __init__(
        self,
        hass: HomeAssistant,
        watersmart: WaterSmartClient,
        hostname: str,
        username: str,
    ) -> None:
        """Initialize."""

        super().__init__(
            hass,
            _LOGGER,
            name=f"WaterSmart {hostname}",
            update_interval=DEFAULT_SCAN_INTERVAL,
        )

        self.watersmart = watersmart
        self.hostname = hostname
        self.username = username
        self.device_info = _get_device_info(hostname, username)
        self.data: CoordinatorData = {}
        self.data_converters = (
            _sensor_data_for_most_recent_hour,
            _sensor_data_for_most_recent_full_day,
            _sensor_data_for_today,
        )

    async def _async_update_data(self) -> CoordinatorData:
        """Update data via library.

        Returns:
            The updated data.

        Raises:
            UpdateFailed: If there is an error that could typically occur.
        """
        try:
            async with timeout(30):
                result: CoordinatorData = {
                    "hourly": await self.watersmart.async_get_hourly_data(),
                }
        except EXCEPTIONS as error:
            raise UpdateFailed(error) from error

        for converter in self.data_converters:
            cast("dict[str, SensorData]", result)[converter.converter_key] = converter(
                result
            )

        _LOGGER.debug("Async update complete")

        return result


def _get_device_info(hostname: str, username: str) -> DeviceInfo:
    """Get device info.

    Returns:
        The device info.
    """
    return DeviceInfo(
        entry_type=DeviceEntryType.SERVICE,
        identifiers={(DOMAIN, f"{hostname}-{username}")},
        manufacturer=MANUFACTURER,
        name=f"WaterSmart ({hostname})",
    )


def _from_timestamp(timestamp: int) -> dt.datetime:
    return dt.datetime.fromtimestamp(timestamp, tz=dt.UTC).replace(
        tzinfo=get_default_time_zone()
    )


class _DataConverter:
    def __init__(
        self,
        key: SensorKey,
        func: Callable[[CoordinatorData], SensorData],
    ) -> None:
        super().__init__()
        self.converter_key = key
        self.func = func

    def __call__(self, data: CoordinatorData) -> SensorData:
        return self.func(data)


def _data_converter[F: Callable[..., Any]](
    key: SensorKey,
) -> Callable[[F], _DataConverterT]:
    """Annotate and add a converter key to data converters.

    Returns:
        A decorator.
    """

    def wrapper(func: F) -> _DataConverterT:
        return cast(
            "_DataConverter",
            functools.wraps(func)(_DataConverter(key, func)),
        )

    return wrapper


@_data_converter(SensorKey.GALLONS_FOR_MOST_RECENT_HOUR)
def _sensor_data_for_most_recent_hour(data: CoordinatorData) -> SensorData:
    """Extract data for most recent hour.

    Returns:
        The extracted & converted records.
    """

    records = data["hourly"][-24:]
    record = records[-1]
    record_date = as_local(_from_timestamp(record["read_datetime"]))

    return {
        "state": _record_gallons(record),
        "attrs": {
            "start": record_date.isoformat(),
            "related": _serialize_records(records),
        },
    }


@_data_converter(SensorKey.GALLONS_FOR_MOST_RECENT_FULL_DAY_KEY)
def _sensor_data_for_most_recent_full_day(data: CoordinatorData) -> SensorData:
    """Extract data for first full day.

    Returns:
        The extracted & converted records.
    """

    records = _records_from_first_full_day(data)
    gallons = sum(_record_gallons(r) for r in records)

    return {
        "state": gallons,
        "attrs": {
            "related": _serialize_records(records),
        },
    }


@_data_converter(SensorKey.GALLONS_TODAY)
def _sensor_data_for_today(data: CoordinatorData) -> SensorData:
    """Extract data for today.

    Returns:
        The extracted & converted records.
    """

    records = _records_from_today(data)
    gallons = sum(_record_gallons(r) for r in records)

    return {
        "state": gallons,
        "attrs": {
            "related": _serialize_records(records),
        },
    }


def _records_from_today(data: CoordinatorData) -> list[UsageRecord]:
    """Extract records for today.

    Returns:
        The extracted records.
    """

    today_records = []
    start_of_day = start_of_local_day()

    for record in reversed(data["hourly"]):
        record_date = as_local(_from_timestamp(record["read_datetime"]))
        if record_date >= start_of_day:
            today_records.append(record)
        else:
            break

    return list(reversed(today_records))


def _records_from_first_full_day(data: CoordinatorData) -> list[UsageRecord]:
    """Extract records for first full day.

    Returns:
        The extracted records.
    """

    full_day_records = []
    last_full_day = None
    start_of_today = start_of_local_day()

    for record in reversed(data["hourly"]):
        record_date = as_local(_from_timestamp(record["read_datetime"]))
        start_of_day = start_of_local_day(record_date)

        if start_of_day >= start_of_today:
            continue

        if last_full_day and start_of_day < last_full_day:
            break

        if last_full_day and start_of_day == last_full_day:
            full_day_records.append(record)
        elif (
            not last_full_day
            and (record_date - start_of_day).total_seconds() // 3600 >= 23
        ):
            full_day_records.append(record)
            last_full_day = start_of_day

    return list(reversed(full_day_records))


def _record_gallons(record: UsageRecord) -> float | int:
    """Get record gallons guarded to ensure it's a number.

    Returns:
        The gallons or zero if it was not available.
    """

    result = record["gallons"]
    if result is None:
        result = 0
    return result


def _serialize_records(records: list[UsageRecord]) -> list[dict[str, Any]]:
    """Convert records for returning in attributes & service calls.

    Returns:
        The serialized records.
    """

    return [
        {
            "start": as_local(_from_timestamp(record["read_datetime"])).isoformat(),
            "gallons": _record_gallons(record),
        }
        for record in records
    ]
