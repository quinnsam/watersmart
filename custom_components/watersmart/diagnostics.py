"""Diagnostics support for WaterSmart."""

from typing import Any, cast

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
from homeassistant.core import HomeAssistant

from .coordinator import WaterSmartUpdateCoordinator

TO_REDACT = {
    CONF_USERNAME,
    CONF_PASSWORD,
}


async def async_get_config_entry_diagnostics(  # noqa: RUF029
    hass: HomeAssistant,  # noqa: ARG001
    entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator: WaterSmartUpdateCoordinator = entry.runtime_data.coordinator

    return cast(
        "dict[str, Any]",
        async_redact_data(
            {
                "entry": entry.as_dict(),
                "data": coordinator.data,
            },
            TO_REDACT,
        ),
    )
