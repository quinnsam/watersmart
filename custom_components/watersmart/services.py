"""Support for the WaterSmart integration."""

from datetime import date, datetime
from functools import partial
from typing import Final, cast

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.core import (
    HomeAssistant,
    ServiceCall,
    ServiceResponse,
    SupportsResponse,
    callback,
)
from homeassistant.exceptions import ServiceValidationError
from homeassistant.helpers import selector
from homeassistant.util import dt
import voluptuous as vol

from .const import DOMAIN
from .coordinator import (
    WaterSmartUpdateCoordinator,
    _from_timestamp,
    _serialize_records,
)
from .types import WaterSmartData

ATTR_CONFIG_ENTRY: Final = "config_entry"
ATTR_FROM_CACHE: Final = "cached"
ATTR_START: Final = "start"
ATTR_END: Final = "end"
HOURLY_HISTORY_SERVICE_NAME: Final = "get_hourly_history"

SERVICE_SCHEMA: Final = vol.Schema(
    {
        vol.Required(ATTR_CONFIG_ENTRY): selector.ConfigEntrySelector(
            {
                "integration": DOMAIN,
            }
        ),
        vol.Optional(ATTR_FROM_CACHE): bool,
        vol.Optional(ATTR_START): vol.Any(str, int),
        vol.Optional(ATTR_END): vol.Any(str, int),
    }
)


def __get_date(date_input: str | int | None) -> date | datetime | None:
    """Get date."""

    if not date_input:
        return None

    if isinstance(date_input, int) and (value := dt.utc_from_timestamp(date_input)):
        return cast(datetime, value)

    if isinstance(date_input, str) and (value := dt.parse_datetime(date_input)):
        return cast(datetime, value)

    raise ServiceValidationError(
        translation_domain=DOMAIN,
        translation_key="invalid_date",
        translation_placeholders={
            "date": date_input,
        },
    )


def __get_coordinator(
    hass: HomeAssistant, call: ServiceCall
) -> WaterSmartUpdateCoordinator:
    """Get the coordinator from the entry."""

    entry_id: str = call.data[ATTR_CONFIG_ENTRY]
    entry: ConfigEntry | None = hass.config_entries.async_get_entry(entry_id)

    if not entry:
        raise ServiceValidationError(
            translation_domain=DOMAIN,
            translation_key="invalid_config_entry",
            translation_placeholders={
                "config_entry": entry_id,
            },
        )

    if entry.state != ConfigEntryState.LOADED:
        raise ServiceValidationError(
            translation_domain=DOMAIN,
            translation_key="unloaded_config_entry",
            translation_placeholders={
                "config_entry": entry.title,
            },
        )

    runtime_data: WaterSmartData = hass.data[DOMAIN][entry_id]

    return runtime_data.coordinator


async def __get_hourly_history(
    call: ServiceCall,
    *,
    hass: HomeAssistant,
) -> ServiceResponse:
    coordinator = __get_coordinator(hass, call)

    start = __get_date(call.data.get(ATTR_START))
    end = __get_date(call.data.get(ATTR_END))

    if call.data.get(ATTR_FROM_CACHE) is False:
        await coordinator.async_refresh()

    records = []

    for record in coordinator.data["hourly"]:
        record_date = _from_timestamp(record["read_datetime"])

        if start and dt.as_local(record_date) < dt.as_local(start):
            continue

        if end and dt.as_local(record_date) > dt.as_local(end):
            continue

        records.append(record)

    return {"history": _serialize_records(records)}


@callback
def async_setup_services(hass: HomeAssistant) -> None:
    """Set up WaterSmart services."""

    hass.services.async_register(
        DOMAIN,
        HOURLY_HISTORY_SERVICE_NAME,
        partial(__get_hourly_history, hass=hass),
        schema=SERVICE_SCHEMA,
        supports_response=SupportsResponse.ONLY,
    )
