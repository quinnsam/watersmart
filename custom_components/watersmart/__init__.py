"""The WaterSmart integration."""

from __future__ import annotations

from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME, Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.typing import ConfigType

from .client import WaterSmartClient
from .const import DOMAIN
from .coordinator import WaterSmartUpdateCoordinator
from .services import async_setup_services
from .types import WaterSmartConfigEntry, WaterSmartData

PLATFORMS: list[Platform] = [Platform.SENSOR]
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up WaterSmart services."""

    async_setup_services(hass)

    return True


async def async_setup_entry(hass: HomeAssistant, entry: WaterSmartConfigEntry) -> bool:
    """Set up WaterSmart from a config entry."""

    hostname: str = entry.data[CONF_HOST]
    username: str = entry.data[CONF_USERNAME]
    password: str = entry.data[CONF_PASSWORD]

    session = async_get_clientsession(hass)
    watersmart = WaterSmartClient(hostname, username, password, session=session)

    coordinator = WaterSmartUpdateCoordinator(
        hass,
        watersmart,
        hostname,
        username,
        password,
    )

    await coordinator.async_config_entry_first_refresh()

    entry.runtime_data = WaterSmartData(
        coordinator=coordinator,
    )

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = entry.runtime_data

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: WaterSmartConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
