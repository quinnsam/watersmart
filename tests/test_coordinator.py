"""Test the WaterSmart coordinator."""

import datetime as dt

from freezegun import freeze_time
from homeassistant.util.dt import get_default_time_zone, start_of_local_day

from custom_components.watersmart.coordinator import WaterSmartUpdateCoordinator


@freeze_time("2024-07-18 12:00:00-07:00")
async def test_records_from_first_full_day_continue_today(
    hass, mock_watersmart_client, snapshot
):
    """Test _records_from_first_full_day to cover the continue condition for today's records."""
    hourly = []
    now = dt.datetime.now(tz=get_default_time_zone())

    # Records for today (should trigger continue)
    hourly.extend(
        [
            {
                "read_datetime": int((now - dt.timedelta(hours=i)).timestamp()),
                "gallons": 1.0,
                "read_id": i,
                "meter_id": "meter1",
            }
            for i in range(5)
        ]
    )

    # Records for a past full day (to ensure the function eventually finds a full day)
    past_day_start = start_of_local_day(now) - dt.timedelta(days=2)
    hourly.extend(
        [
            {
                "read_datetime": int(
                    (past_day_start + dt.timedelta(hours=i)).timestamp()
                ),
                "gallons": 2.0,
                "read_id": 100 + i,
                "meter_id": "meter1",
            }
            for i in range(24)
        ]
    )

    hourly.reverse()  # Simulate chronological order from client

    mock_watersmart_client.async_get_hourly_data.return_value = hourly

    coordinator = WaterSmartUpdateCoordinator(
        hass,
        mock_watersmart_client,
        "test_host",
        "test_user",
    )

    data = await coordinator._async_update_data()

    assert snapshot == data
