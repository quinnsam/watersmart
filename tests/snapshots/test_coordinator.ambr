# serializer version: 1
# name: test_records_from_first_full_day_continue_today
  dict({
    <SensorKey.GALLONS_FOR_MOST_RECENT_FULL_DAY_KEY: 'gallons_for_most_recent_full_day_key'>: dict({
      'attrs': dict({
        'related': list([
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T23:00:00-07:00',
          }),
        ]),
      }),
      'state': 2.0,
    }),
    <SensorKey.GALLONS_FOR_MOST_RECENT_HOUR: 'gallons_for_most_recent_hour'>: dict({
      'attrs': dict({
        'related': list([
          dict({
            'gallons': 2.0,
            'start': '2024-07-17T01:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-17T00:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T23:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T22:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T21:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T20:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T19:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T18:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T17:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T16:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T15:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T14:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T13:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T12:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T11:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T10:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T09:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T08:00:00-07:00',
          }),
          dict({
            'gallons': 2.0,
            'start': '2024-07-16T07:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T15:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T16:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T17:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T18:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T19:00:00-07:00',
          }),
        ]),
        'start': '2024-07-18T19:00:00-07:00',
      }),
      'state': 1.0,
    }),
    <SensorKey.GALLONS_TODAY: 'gallons_today'>: dict({
      'attrs': dict({
        'related': list([
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T15:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T16:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T17:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T18:00:00-07:00',
          }),
          dict({
            'gallons': 1.0,
            'start': '2024-07-18T19:00:00-07:00',
          }),
        ]),
      }),
      'state': 5.0,
    }),
    'hourly': list([
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721196000,
        'read_id': 123,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721192400,
        'read_id': 122,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721188800,
        'read_id': 121,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721185200,
        'read_id': 120,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721181600,
        'read_id': 119,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721178000,
        'read_id': 118,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721174400,
        'read_id': 117,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721170800,
        'read_id': 116,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721167200,
        'read_id': 115,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721163600,
        'read_id': 114,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721160000,
        'read_id': 113,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721156400,
        'read_id': 112,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721152800,
        'read_id': 111,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721149200,
        'read_id': 110,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721145600,
        'read_id': 109,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721142000,
        'read_id': 108,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721138400,
        'read_id': 107,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721134800,
        'read_id': 106,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721131200,
        'read_id': 105,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721127600,
        'read_id': 104,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721124000,
        'read_id': 103,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721120400,
        'read_id': 102,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721116800,
        'read_id': 101,
      }),
      dict({
        'gallons': 2.0,
        'meter_id': 'meter1',
        'read_datetime': 1721113200,
        'read_id': 100,
      }),
      dict({
        'gallons': 1.0,
        'meter_id': 'meter1',
        'read_datetime': 1721314800,
        'read_id': 4,
      }),
      dict({
        'gallons': 1.0,
        'meter_id': 'meter1',
        'read_datetime': 1721318400,
        'read_id': 3,
      }),
      dict({
        'gallons': 1.0,
        'meter_id': 'meter1',
        'read_datetime': 1721322000,
        'read_id': 2,
      }),
      dict({
        'gallons': 1.0,
        'meter_id': 'meter1',
        'read_datetime': 1721325600,
        'read_id': 1,
      }),
      dict({
        'gallons': 1.0,
        'meter_id': 'meter1',
        'read_datetime': 1721329200,
        'read_id': 0,
      }),
    ]),
  })
# ---
